orangecontrib/spectroscopy/widgets/owpolar.py

```diff
@@ -0,0 +1,823 @@
+
+        for t in threads:
+            t.join()
+
+        state.set_status("Finishing...")
+        if invert is True:
+            sout[:,:,:,2] = sout[:,:,:,2]*-1
+        outputs = np.reshape(sout[:,:,:,2:], (np.shape(ulsys)[0]*np.shape(ulsxs)[0], 5*len(featnames)))
+        model = np.reshape(smod[:,:,:,2:], (np.shape(ulsys)[0]*np.shape(ulsxs)[0], 4*len(featnames)))
+
+    spectra = []
+    met = []
+    if average is False:
+        for i in range(len(polangles)):
+            spectratemp = np.reshape(spec[:,:,:,i],
+                                    (np.shape(ulsys)[0]*np.shape(ulsxs)[0], images[0].X.shape[1]))
+            spectratemp = spectratemp[~np.isnan(model).any(axis=1)]
+            spectra.append(spectratemp)
+            metatemp = np.reshape(metas[:,:,:,i],
+                                (np.shape(ulsys)[0]*np.shape(ulsxs)[0], images[0].metas.shape[1]))
+            metatemp = metatemp[~np.isnan(model).any(axis=1)]
+            metatemp = np.append(metatemp, np.full((np.shape(metatemp)[0],1), i), axis=1)
+            met.append(metatemp)
+    elif average is True:
+        average_spec = np.average(spec, axis=3)
+        spectratemp = np.reshape(average_spec,
+                                (np.shape(ulsys)[0]*np.shape(ulsxs)[0], images[0].X.shape[1]))
+        spectratemp = spectratemp[~np.isnan(model).any(axis=1)]
+        spectra.append(spectratemp)
+        metatemp = np.reshape(metas[:,:,:,0],
+                            (np.shape(ulsys)[0]*np.shape(ulsxs)[0], images[0].metas.shape[1]))
+        metatemp = metatemp[~np.isnan(model).any(axis=1)]
+        met.append(metatemp)
+
+    outputs = outputs[~np.isnan(model).any(axis=1)]
+    model = model[~np.isnan(model).any(axis=1)]
+
+    spectra = np.concatenate((spectra), axis=0)
+    meta = np.concatenate((met), axis=0)
+
+    return outputs, model, spectra, meta, vars[1]
+
+
+class OWPolar(OWWidget, ConcurrentWidgetMixin):
+
+    # Widget's name as displayed in the canvas
+    name = "4+ Angle Polarisation"
+
+    # Short widget description
+    description = (
+        "Calculate Azimuth Angle, Orientation function, Amplitude and Intensity of "
+        "vibrational mode(s) using polarised data measured at 4 or more polarisation angles.")
+
+    icon = "icons/polar.svg"
+
+    # Define inputs and outputs
+    class Inputs:
+        data = MultiInput("Data", Orange.data.Table, default=True)
+
+    class Outputs:
+        polar = Output("Polar Data", Orange.data.Table, default=True)
+        model = Output("Curve Fit model data", Orange.data.Table)
+
+    autocommit = settings.Setting(False)
+
+    settingsHandler = DomainContextHandler()
+
+    want_main_area = False
+    resizing_enabled = True
+    alpha = Setting(0, schema_only=True)
+
+    map_x = ContextSetting(None)
+    map_y = ContextSetting(None)
+    invert_angles = Setting(False, schema_only=True)
+    average = Setting(False, schema_only=True)
+    angles = ContextSetting(None)
+
+    anglst = []
```

Comment:
The object variables need to be initialized in `__init__`. As they are currently defined, they are class variables. Instances of the widget can therefore share them (it depends on the specifics, but is highly dangerous. The code extract below showcases the potential issues.

Orange's Settings seem like they do not follow this rule, but they are magic. :) When the class is initialized, whatever is a `Setting` is handled in a special way. Therefore, it is fine to have settings initialized like this, but not ordinary variables.

```
class A:
    i = 1
    l = []

a1 = A()
a2 = A()

a1.i = 2
print(a2.i)

a1.l.append("whoooops")
print(a2.l)
```

This yields:
```
1
['whoooops']
```

Reply:
Fixed in today's commits.